cpp/src/copying/gather.cu

```diff
@@ -0,0 +1,396 @@
+/*
+ * Copyright (c) 2018, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cooperative_groups.h>
+#include <thrust/gather.h>
+#include "copying.hpp"
+#include "cudf.h"
+#include "gather.hpp"
+#include "rmm/thrust_rmm_allocator.h"
+#include "utilities/cudf_utils.h"
+#include "utilities/type_dispatcher.hpp"
+
+/**
+ * @brief Operations for copying from one column to another
+ * @file copying_ops.cu
+ */
+
+namespace {
+
+/**---------------------------------------------------------------------------*
+ * @brief Function object to check if an index is within the bounds [begin,
+ * end).
+ *
+ *---------------------------------------------------------------------------**/
+struct bounds_checker {
+  gdf_index_type const begin;
+  gdf_index_type const end;
+
+  __device__ bounds_checker(gdf_index_type begin_, gdf_index_type end_)
+      : begin{begin_}, end{end_} {}
+
+  __device__ __forceinline__ bool operator()(gdf_index_type const index) {
```

Comment:
The `__forceinline_` here is completely redundant and unnecessary anyways because it's a member function of a `struct` that's defined inline, it will already be inlined... just a force of habit.